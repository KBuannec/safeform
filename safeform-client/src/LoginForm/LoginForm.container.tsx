import LoginForm from './LoginForm';
import { useLoginForm } from './LoginForm.hook';

const LoginFormContainer = () => {
    const { form, errors, success, handleChange, handleSubmit } = useLoginForm();

    return (
        <div className="register-form-wrapper">
            <LoginForm
                form={form}
                errors={errors}
                success={success}
                onChange={handleChange}
                onSubmit={handleSubmit}
            />
        </div>
    );
};

export default LoginFormContainer;
