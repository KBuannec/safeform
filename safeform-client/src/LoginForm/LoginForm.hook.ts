import { useEffect, useState } from 'react';
import { z } from 'zod';

const loginSchema = z.object({
  email: z.string().email('Email invalide'),
  password: z.string().min(6, 'Mot de passe requis'),
});

export type LoginFormData = z.infer<typeof loginSchema>;

export const useLoginForm = () => {
  const [form, setForm] = useState<LoginFormData>({ email: '', password: '' });
  const [errors, setErrors] = useState<Partial<Record<keyof LoginFormData, string>>>({});
  const [success, setSuccess] = useState<string | null>(null);
  const [csrfToken, setCsrfToken] = useState<string>('');

  useEffect(() => {
    fetch('http://localhost:3000/csrf-token', {
      credentials: 'include',
    })
      .then(res => res.json() as Promise<{ csrfToken: string }>)
      .then(data => {
        setCsrfToken(data.csrfToken);
        console.log('✅ Token CSRF (login) :', data.csrfToken);
      })
      .catch(() => {
        console.warn('Échec récupération token CSRF (login)');
      });
  }, []);

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setForm({ ...form, [e.target.name]: e.target.value });
  };

  const handleSubmit = async (e: React.FormEvent) => {
    e.preventDefault();

    const result = loginSchema.safeParse(form);

    if (!result.success) {
      const fieldErrors = result.error.flatten().fieldErrors;
      setErrors({
        email: fieldErrors.email?.[0],
        password: fieldErrors.password?.[0],
      });
      return;
    }

    setErrors({});
    try {
      const res = await fetch('http://localhost:3000/login', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-XSRF-TOKEN': csrfToken,
        },
        credentials: 'include',
        body: JSON.stringify(form),
      });

      const data = await res.json();

      if (!res.ok) {
        setSuccess(null);
        setErrors({ email: data.message || 'Identifiants incorrects' });
        return;
      }

      setSuccess('Connexion réussie !');
      setForm({ email: '', password: '' });
    } catch (err) {
      console.error(err);
      setSuccess("Une erreur est survenue");
    }
  };

  return {
    form,
    errors,
    success,
    handleChange,
    handleSubmit,
  };
};
