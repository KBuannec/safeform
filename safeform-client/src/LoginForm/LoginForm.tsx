import { useState } from 'react';
import './LoginForm.scss';

interface LoginFormProps {
  form: {
    email: string;
    password: string;
  };
  errors: {
    email?: string;
    password?: string;
  };
  success: string | null;
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  onSubmit: React.FormEventHandler<HTMLFormElement>;
}

const LoginForm = ({ form, errors, success, onChange, onSubmit }: LoginFormProps) => {
  const [showPassword, setShowPassword] = useState(false);

  return (
    <form className='register-form' onSubmit={onSubmit}>
      <h2>Connexion</h2>

      <label htmlFor='email'>Email</label>
      <input id='email' name='email' value={form.email} onChange={onChange} placeholder='Email' />
      {errors.email && <p className="error">{errors.email}</p>}

      <label htmlFor='password'>Mot de passe</label>
      <div className="input-wrapper">
        <input
          id='password'
          name='password'
          type={showPassword ? 'text' : 'password'}
          value={form.password}
          onChange={onChange}
          placeholder='••••••••'
        />
        <span className='eye-icon' onClick={() => setShowPassword(p => !p)}>
          {showPassword ? '🙈' : '👁️'}
        </span>
      </div>
      {errors.password && <p className="error">{errors.password}</p>}

      <button type='submit'>Se connecter</button>

      {success && <p className='success'>{success}</p>}
    </form>
  );
};

export default LoginForm;
