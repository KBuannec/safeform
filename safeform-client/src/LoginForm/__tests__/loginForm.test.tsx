import { describe, it, expect, vi } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import LoginForm from '../LoginForm';

describe('LoginForm component', () => {
    const baseProps = {
        form: {
            email: '',
            password: '',
        },
        errors: {},
        success: null,
        onChange: vi.fn(),
        onSubmit: vi.fn(),
    };

    it('affiche le titre de connexion', () => {
        render(<LoginForm { ...baseProps } />);
        expect(screen.getByRole('heading', { name: /connexion/i })).toBeInTheDocument();
    });

    it('affiche les champs email et mot de passe', () => {
        render(<LoginForm { ...baseProps } />);
        expect(screen.getByPlaceholderText('Email')).toBeInTheDocument();
        expect(screen.getByPlaceholderText('••••••••')).toBeInTheDocument();
    });

    it('affiche les erreurs si présentes', () => {
        render(
            <LoginForm
        { ...baseProps }
        errors = {{ email: 'Email invalide', password: 'Mot de passe requis' }}
      />
    );
    expect(screen.getByText('Email invalide')).toBeInTheDocument();
    expect(screen.getByText('Mot de passe requis')).toBeInTheDocument();
});

it('bascule la visibilité du mot de passe au clic sur l’icône', () => {
    render(<LoginForm { ...baseProps } />);
    const toggleIcon = screen.getByText('👁️');
    fireEvent.click(toggleIcon);
    expect(screen.getByPlaceholderText('••••••••')).toHaveAttribute('type', 'text');
});

it('correspond au snapshot', () => {
    const { asFragment } = render(<LoginForm { ...baseProps } />);
    expect(asFragment()).toMatchSnapshot();
});
});
