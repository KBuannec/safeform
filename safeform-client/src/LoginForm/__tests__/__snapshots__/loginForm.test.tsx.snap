// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginForm component > correspond au snapshot 1`] = `
<DocumentFragment>
  <form
    class="login-form"
  >
    <h2>
      Connexion
    </h2>
    <label
      for="email"
    >
      Email
    </label>
    <input
      id="email"
      name="email"
      placeholder="Email"
      value=""
    />
    <label
      for="password"
    >
      Mot de passe
    </label>
    <div
      class="input-wrapper"
    >
      <input
        id="password"
        name="password"
        placeholder="••••••••"
        type="password"
        value=""
      />
      <span
        class="eye-icon"
      >
        👁️
      </span>
    </div>
    <button
      type="submit"
    >
      Se connecter
    </button>
  </form>
</DocumentFragment>
`;
